Suggestion on rate-limited (#3)
mwelzl/draft-ccwg-constrained-increase: draft-welzl-ccwg-constrained-increase.md

--- abstract

This document specifies how transport protocols should increase their congestion window when the sender is constrained.
Such a constraint can  be caused by the application stopping to supply data or by flow control.

--- middle

# Introduction

We call a sender of a congestion controlled transport protocol "constrained" when it does not have data to send (i.e., the application has not provided sufficient data) even though the congestion control rules would allow it to transmit data.
A sender can also be constrained by the receiver when flow control limits apply (e.g., via the receiver window (rwnd) in case of TCP, of the flow credit in quic).
RFCs specifying congestion control mechanisms diverge regarding the rules for increasing the congestion window (cwnd) when the sender is constrained.

> This is also known as a rate-limited sender.

Reviewer: ```suggestion
```
I take this out because I'll instead change all the terminology to rate-limited afterwards.

cwnd = min(cwnd_new, 1+maxFS)
~~~

maxFS is the largest FlightSize value since the last time that cwnd was decreased.
If cwnd has never been decreased, it is the maximum FlightSize value since the beginning of the data transfer.

## Discussion

> If the sender rate is limited for multiple RTTs, either by the sending application or by the receivin applicatiom, continuously increasing the cwnd would cause a mismatch between the cwnd and the capacity the path provides (i.e., over-estimating the capacity).

Reviewer: ```suggestion
If the sender rate is limited for multiple RTTs, either by the sending application or by the receiving application, continuously increasing the cwnd would cause a mismatch between the cwnd and the capacity the path provides (i.e., over-estimating the capacity).
```

# The state of RFCs and implementations

This section is provided as input for IETF discussion, and should be removed before publication.

## TCP ("Reno" congestion control)

### Specification

{{!RFC5681}} does not contain a rule to limit the growth of cwnd when the sender is rate-limited. This statement (page 8) gives an impression that such cwnd growth might be expected:

>Implementation Note: An easy mistake to make is to simply use cwnd, rather than FlightSize, which in some implementations may incidentally increase well beyond rwnd.

> {{?RFC7661}} also suggests there is no standard TCP behavior (which {{?RFC7661}} changes), on page 4:

Reviewer: ```suggestion
{{?RFC7661}} also suggests there is no increase limitation in the standard TCP behavior (which {{?RFC7661}} changes), on page 4:
```

### Implementation {#tcp-impl}

> - ns-2 allows cwnd to grow when it is rate-limited by rwnd. (Rate-limited: not tested.)

Reviewer: ```suggestion
- ns-2 allows cwnd to grow when it is rate-limited by rwnd. (Rate-limited by the sending application: not tested.)
```

- ns-3 allows cwnd to grow when it is rate-limited by either an application or the rwnd.
- Linux only allows cwnd to grow when the sender is unconstrained.

>   Specifically, for Linux kernel 6.0.9, an increase is only allowed if a function called `tcp_is_cwnd_limited` in `tcp.h` yields `true`. This function checks the flag `tp->is_cwnd_limited`, which is initialised to `false` in `tcp_output.c` and later set to `true` only if FlightSize is greater or equal to cwnd (`is_cwnd_limited |= (tcp_packets_in_flight(tp) >= tcp_snd_cwnd(tp))`).

Reviewer: ```suggestion
  Specifically, for Linux kernel 2.6.11, an increase is only allowed if a function called `tcp_is_cwnd_limited` in `tcp.h` yields `true`. This function checks the flag `tp->is_cwnd_limited`, which is initialised to `false` in `tcp_output.c` and later set to `true` only if FlightSize is greater or equal to cwnd (`is_cwnd_limited |= (tcp_packets_in_flight(tp) >= tcp_snd_cwnd(tp))`).
```

## SCTP

### Specification

{{Section 7.2.1 of !RFC9260}} says:

>When cwnd is less than or equal to ssthresh, an SCTP endpoint MUST use the slow-start algorithm to increase cwnd only if the current congestion window is being fully utilized and the data sender is not in Fast Recovery.
Only when these two conditions are met can the cwnd be increased; otherwise, the cwnd MUST NOT be increased.

### Assessment

The quoted statement from {{!RFC9260}} prescribes the same cwnd growth limitation that is also specified for Cubic and implemented for both Reno and Cubic in Linux. It is in accordance with rule #1 in {{rules}}, and more conservative than rule #2 in {{rules}}.

{{Section 7.2.1 of !RFC9260}} is specifically limited to Slow Start. Congestion Avoidance is discussed in {{Section 7.2.2 of !RFC9260}}, but this section neither contains a similar rule nor does it refer back to the rule that limits the growth of cwnd
in Section 7.2.1. It is thus implicitly clear that the quoted rule only applies to Slow Start, whereas the rules in {{rules}} apply to both Slow Start and Congestion Avoidance.

## QUIC

### Specification

{{Section 7.8 of !RFC9002}} states:

> >When bytes in flight is smaller than the congestion window and sending is not pacing-limited, the congestion window is underutilized. This can happen due to insufficient application data or flow control limits. When this occurs, the congestion window SHOULD NOT be increased in either slow start or congestion avoidance.

Reviewer: ```suggestion
>When bytes in flight is smaller than the congestion window and sending is not pacing limited, the congestion window is underutilized. This can happen due to insufficient application data or flow control limits. When this occurs, the congestion window SHOULD NOT be increased in either slow start or congestion avoidance.
```

Undoing this change because it's a literal quote from RFC 9002

>A sender that paces packets (see Section 7.7) might delay sending packets and not fully utilize the congestion window due to this delay. A sender SHOULD NOT consider itself application limited if it would have fully utilized the congestion window without pacing delay.

### Assessment

> With the exception of pacing, this specification conservatively limits the grworth in cwnd, similar to Cubic and SCTP. The exception for pacing in the second paragraph requires the application to notify the transport layer that it paces packets. Pacing is typically done with delays below an RTT; thus, rule #2 in {{rules}} should cover this case without the need for such a notification from the application.

Reviewer: ```suggestion
With the exception of pacing, this specification conservatively limits the growth in cwnd, similar to Cubic and SCTP. The exception for pacing in the second paragraph requires the application to notify the transport layer that it paces packets. Pacing is typically done with delays below an RTT; thus, rule #2 in {{rules}} should cover this case without the need for such a notification from the application.
```
